File: mac/HockeyKit/Sources/HockeyCore/Quarters.swift
/// Quarters: spans the coach marks on the game timeline, the quarter clock
/// read from them, the timeline bands, and the breaks playback skips.
///
/// Ports of `src/features/quarters/navigation.ts`, `clock.ts`, `validation.ts`
/// and `draft.ts`, pinned by `contracts/vectors/quarters.json` and
/// `quarter-draft.json`. A quarter is half-open `[startS, end)`; an unset end
/// runs to the next quarter's start. How many periods a game has and how long
/// each lasts are the game's format (`GameFormat`), an input to every rule
/// here.

/// A quarter on the game timeline; `index` counts from 1.
public struct Quarter: Equatable, Hashable, Sendable {
    public let index: Int
    public let startS: Double
    /// The marked end, or `nil` when only the start is marked.
    public let endS: Double?

    public init(index: Int, startS: Double, endS: Double?) {
        self.index = index
        self.startS = startS
        self.endS = endS
    }
}

/// A quarter's span as fractions of the game, for the timeline bands.
public struct QuarterBand: Equatable, Sendable {
    public let index: Int
    public let startFraction: Double
    public let endFraction: Double
}

/// The quarters ordered by index; equal indices keep their order.
private func byIndex(_ quarters: [Quarter]) -> [Quarter] {
    quarters.enumerated()
        .sorted { lhs, rhs in
            lhs.element.index != rhs.element.index
                ? lhs.element.index < rhs.element.index : lhs.offset < rhs.offset
        }
        .map(\.element)
}

/// The effective end of quarter `i` of an index-sorted list: its marked end,
/// else the next start, else `fallbackS`.
private func effectiveEndS(_ sorted: [Quarter], _ i: Int, fallbackS: Double) -> Double {
    if let endS = sorted[i].endS { return endS }
    return i + 1 < sorted.count ? sorted[i + 1].startS : fallbackS
}

/// The quarter holding `gameTimeS`, or `nil` in a break, before the first or
/// after a marked last end.
public func quarterAt(_ quarters: [Quarter], gameTimeS: Double) -> Quarter? {
    guard gameTimeS.isFinite else { return nil }
    let sorted = byIndex(quarters)
    for (i, quarter) in sorted.enumerated() where gameTimeS >= quarter.startS {
        if gameTimeS < effectiveEndS(sorted, i, fallbackS: .infinity) { return quarter }
    }
    return nil
}

/// A quarter's clip window, clamped to the game; `nil` when no quarter has
/// that index.
public func quarterWindow(
    _ quarters: [Quarter],
    index: Int,
    totalDurationS: Double
) -> (startS: Double, endS: Double)? {
    let sorted = byIndex(quarters)
    guard let i = sorted.firstIndex(where: { $0.index == index }) else { return nil }
    let startS = max(0, min(sorted[i].startS, totalDurationS))
    let endS = min(effectiveEndS(sorted, i, fallbackS: totalDurationS), totalDurationS)
    return (startS, max(startS, endS))
}

/// The quarters as fractions of the game; none while the game has no length.
public func quarterBands(_ quarters: [Quarter], totalDurationS: Double) -> [QuarterBand] {
    guard totalDurationS > 0 else { return [] }
    let sorted = byIndex(quarters)
    return sorted.indices.map { i in
        QuarterBand(
            index: sorted[i].index,
            startFraction: clampFraction(sorted[i].startS / totalDurationS),
            endFraction: clampFraction(effectiveEndS(sorted, i, fallbackS: totalDurationS) / totalDurationS)
        )
    }
}

/// Where playback jumps to skip a break: the next quarter's start while
/// `gameTimeS` lies between a marked end and that start, else `nil`. Footage
/// before the first quarter and after the last is never skipped.
public func breakSkipTargetS(_ quarters: [Quarter], gameTimeS: Double) -> Double? {
    guard gameTimeS.isFinite else { return nil }
    let sorted = byIndex(quarters)
    for i in sorted.indices.dropLast() {
        guard let endS = sorted[i].endS else { continue }
        let nextStartS = sorted[i + 1].startS
        if gameTimeS >= endS, gameTimeS < nextStartS { return nextStartS }
    }
    return nil
}

/// The quarter clock: inside quarter `i` it reads `(i - 1) * periodLengthS`
/// plus the time since the quarter's start; outside every quarter the raw game
/// time runs on.
public func quarterClockS(_ quarters: [Quarter], gameTimeS: Double, periodLengthS: Double) -> Double {
    guard let quarter = quarterAt(quarters, gameTimeS: gameTimeS) else { return gameTimeS }
    return Double(quarter.index - 1) * periodLengthS + (gameTimeS - quarter.startS)
}

private func clampFraction(_ fraction: Double) -> Double {
    fraction.isFinite ? min(max(fraction, 0), 1) : 0
}

// MARK: - Validation

/// Why a quarter set cannot be stored.
public enum QuartersError: Error, Equatable, Sendable {
    /// The set is empty or has more quarters than the game's format.
    case count
    /// An index lies outside `1...periodCount`, a start is negative or not
    /// finite, or an end does not lie after its start.
    case invalidQuarter
    /// The indices do not run 1, 2, 3, ... without a gap or a repeat.
    case notContiguous
    /// A quarter starts at or before the previous one.
    case order
    /// A quarter starts before the previous quarter's marked end.
    case overlap
}

/// Checks a quarter set before it is stored and returns it sorted by index:
/// at least one and at most `periodCount` quarters, indices `1...N`, each
/// start after the previous, no overlap.
public func validateQuarters(_ quarters: [Quarter], periodCount: Int) throws(QuartersError) -> [Quarter] {
    guard !quarters.isEmpty, quarters.count <= periodCount else { throw .count }
    for quarter in quarters {
        guard (1...max(periodCount, 1)).contains(quarter.index),
              quarter.startS.isFinite, quarter.startS >= 0
        else { throw .invalidQuarter }
        if let endS = quarter.endS, !endS.isFinite || endS <= quarter.startS {
            throw .invalidQuarter
        }
    }
    let sorted = byIndex(quarters)
    for (i, quarter) in sorted.enumerated() {
        guard quarter.index == i + 1 else { throw .notContiguous }
        guard i > 0 else { continue }
        let previous = sorted[i - 1]
        if quarter.startS <= previous.startS { throw .order }
        if let endS = previous.endS, endS > quarter.startS { throw .overlap }
    }
    return sorted
}

// MARK: - Editor draft

/// One row of the quarter editor; a `nil` start means not marked yet.
public struct QuarterDraft: Equatable, Sendable, Identifiable {
    public let index: Int
    public var startS: Double?
    public var endS: Double?

    public var id: Int { index }

    public init(index: Int, startS: Double?, endS: Double?) {
        self.index = index
        self.startS = startS
        self.endS = endS
    }
}

/// Why the editor's rows cannot be saved.
public enum QuarterDraftProblem: String, Equatable, Sendable {
    /// A quarter is marked while an earlier one is not.
    case gap
    /// A quarter ends at or before its own start.
    case endBeforeStart
    /// A quarter starts at or before the previous start.
    case order
    /// A quarter starts before the previous quarter's marked end.
    case overlap
}

/// One row per period of the game's format, seeded from the stored quarters.
public func initialDraft(_ quarters: [Quarter], periodCount: Int) -> [QuarterDraft] {
    (1...max(periodCount, 1)).map { index in
        let stored = quarters.first { $0.index == index }
        return QuarterDraft(index: index, startS: stored?.startS, endS: stored?.endS)
    }
}

/// The marked rows in index order: the set that is stored.
public func toQuarters(_ draft: [QuarterDraft]) -> [Quarter] {
    draft.sorted { $0.index < $1.index }.compactMap { row in
        row.startS.map { Quarter(index: row.index, startS: $0, endS: row.endS) }
    }
}

/// The first problem that blocks saving the rows, or `nil`.
public func draftProblem(_ draft: [QuarterDraft]) -> QuarterDraftProblem? {
    let rows = draft.sorted { $0.index < $1.index }
    if let lastMarked = rows.lastIndex(where: { $0.startS != nil }),
       rows[..<lastMarked].contains(where: { $0.startS == nil })
    {
        return .gap
    }
    let quarters = toQuarters(rows)
    for (i, quarter) in quarters.enumerated() {
        if let endS = quarter.endS, endS <= quarter.startS { return .endBeforeStart }
        guard i > 0 else { continue }
        let previous = quarters[i - 1]
        if quarter.startS <= previous.startS { return .order }
        if let endS = previous.endS, endS > quarter.startS { return .overlap }
    }
    return nil
}

File: contracts/generator/quarters.ts
/**
 * Golden vectors for quarters: the validated period set a game stores, and
 * the navigation, band, break-skip and clock rules built on it. Periods are
 * half-open [startS, end) in game time; an unset end runs to the next start.
 * The game's format (see `game-format.ts`) is an input: the clock takes the
 * period length and the validation the period count.
 */
import { DEFAULT_TOLERANCE, vectorCase, type VectorFile } from "./vector";

import { DEFAULT_GAME_FORMAT } from "@/features/game-format/format";
import { quarterClockS } from "@/features/quarters/clock";
import {
  breakSkipTargetS,
  quarterAt,
  quarterBands,
  quarterWindow,
  type Quarter,
} from "@/features/quarters/navigation";
import { parseQuartersInput } from "@/features/quarters/validation";

const GAME_ID = "00000000-0000-4000-8000-000000000001";
const GAME_LENGTH_S = 4800;

/** Four quarters with every end marked, a half-time break and an open last quarter. */
const MARKED: Quarter[] = [
  { index: 1, startS: 120, endS: 1020 },
  { index: 2, startS: 1200, endS: 2100 },
  { index: 3, startS: 2700, endS: 3600 },
  { index: 4, startS: 3780, endS: null },
];

/** Two 20-minute halves (indoor), both ends marked. */
const HALVES: Quarter[] = [
  { index: 1, startS: 90, endS: 1290 },
  { index: 2, startS: 1800, endS: 3000 },
];

/** Four 10-minute quarters (youth), the last one open. */
const SHORT_QUARTERS: Quarter[] = [
  { index: 1, startS: 30, endS: 630 },
  { index: 2, startS: 700, endS: 1300 },
  { index: 3, startS: 1600, endS: 2200 },
  { index: 4, startS: 2300, endS: null },
];

/** Only the starts marked, listed out of order. */
const STARTS_ONLY: Quarter[] = [
  { index: 2, startS: 1000, endS: null },
  { index: 1, startS: 60, endS: null },
];

function atCase(name: string, quarters: Quarter[], gameTimeS: number) {
  return vectorCase(name, "quarterAt", { quarters, gameTimeS }, (i) =>
    quarterAt(i.quarters, i.gameTimeS),
  );
}

function windowCase(
  name: string,
  quarters: Quarter[],
  index: number,
  totalDurationS: number,
) {
  return vectorCase(
    name,
    "quarterWindow",
    { quarters, index, totalDurationS },
    (i) => quarterWindow(i.quarters, i.index, i.totalDurationS),
  );
}

function bandsCase(name: string, quarters: Quarter[], totalDurationS: number) {
  return vectorCase(name, "quarterBands", { quarters, totalDurationS }, (i) =>
    quarterBands(i.quarters, i.totalDurationS),
  );
}

function skipCase(name: string, quarters: Quarter[], gameTimeS: number) {
  return vectorCase(name, "breakSkipTargetS", { quarters, gameTimeS }, (i) =>
    breakSkipTargetS(i.quarters, i.gameTimeS),
  );
}

function clockCase(
  name: string,
  quarters: Quarter[],
  gameTimeS: number,
  periodLengthS: number = DEFAULT_GAME_FORMAT.periodLengthS,
) {
  return vectorCase(
    name,
    "quarterClockS",
    { quarters, gameTimeS, periodLengthS },
    (i) => quarterClockS(i.quarters, i.gameTimeS, i.periodLengthS),
  );
}

function parseCase(
  name: string,
  body: unknown,
  periodCount: number = DEFAULT_GAME_FORMAT.periodCount,
) {
  return vectorCase(name, "parseQuartersInput", { body, periodCount }, (i) => {
    const result = parseQuartersInput(i.body, i.periodCount);
    // The error is English API text; a port rejects the same bodies in its
    // own words, so only the outcome is pinned.
    return result.ok ? result : { ok: false };
  });
}

export function buildQuarters(): VectorFile {
  return {
    contract: "quarters",
    description:
      "Quarter sets and the rules on them. quarterAt finds the quarter holding a " +
      "game time (null in a break); quarterWindow is a quarter's clip window " +
      "clamped to the game; quarterBands are fractions of the game; " +
      "breakSkipTargetS is where a break jumps to; quarterClockS reads " +
      "(index - 1) * periodLengthS plus the time into the period, and raw game " +
      "time outside periods. parseQuartersInput validates a stored set for a " +
      "game playing periodCount periods: contiguous indices from 1 up to " +
      "periodCount, each start after the previous, no overlap (its English " +
      "error is left out). The period length and count are the game's format " +
      "(game-format.json) and always inputs, never assumed.",
    reference: [
      "src/features/quarters/navigation.ts",
      "src/features/quarters/clock.ts",
      "src/features/quarters/validation.ts",
    ],
    tolerance: DEFAULT_TOLERANCE,
    cases: [
      atCase("before the first quarter", MARKED, 30),
      atCase("a quarter's start is inside it", MARKED, 120),
      atCase("inside the first quarter", MARKED, 500),
      atCase("a marked end is already the break", MARKED, 1020),
      atCase("inside a break", MARKED, 1100),
      atCase("the next start ends the break", MARKED, 1200),
      atCase("an open last quarter runs on", MARKED, 6000),
      atCase("an unset end runs to the next start", STARTS_ONLY, 999.98),
      atCase("the next start begins the next quarter", STARTS_ONLY, 1000),
      atCase("no quarters marked", [], 500),

      windowCase("a marked quarter", MARKED, 1, GAME_LENGTH_S),
      windowCase(
        "an open last quarter ends with the game",
        MARKED,
        4,
        GAME_LENGTH_S,
      ),
      windowCase(
        "an unset end is the next start",
        STARTS_ONLY,
        1,
        GAME_LENGTH_S,
      ),
      windowCase(
        "a quarter that is not marked",
        MARKED.slice(0, 2),
        3,
        GAME_LENGTH_S,
      ),
      windowCase("a quarter starting past the game is empty", MARKED, 4, 3700),

      bandsCase("marked quarters", MARKED, GAME_LENGTH_S),
      bandsCase("starts only", STARTS_ONLY, GAME_LENGTH_S),
      bandsCase("no quarters", [], GAME_LENGTH_S),
      bandsCase("a game without length has no bands", MARKED, 0),

      skipCase("a marked end skips to the next start", MARKED, 1020),
      skipCase("inside a break skips to the next start", MARKED, 1100),
      skipCase("one frame before the next start", MARKED, 1199.98),
      skipCase("the next start itself", MARKED, 1200),
      skipCase("inside a quarter", MARKED, 500),
      skipCase("before the first quarter is never skipped", MARKED, 30),
      skipCase("unset ends leave no break", STARTS_ONLY, 999.98),

      clockCase("before the first quarter runs raw", MARKED, 30),
      clockCase("a quarter start reads its nominal start", MARKED, 120),
      clockCase("the clock inside the first quarter", MARKED, 500),
      clockCase("inside the second quarter", MARKED, 1300),
      clockCase("a break runs raw", MARKED, 1100),
      clockCase("the open last quarter", MARKED, 4000),
      clockCase("ten-minute quarters", MARKED, 2800, 600),
      clockCase("a break runs raw whatever the length", MARKED, 1100, 600),
      clockCase("2 x 20: the first half", HALVES, 690, 1200),
      clockCase("2 x 20: the half-time break runs raw", HALVES, 1500, 1200),
      clockCase("2 x 20: the second half reads 20:00", HALVES, 1800, 1200),
      clockCase("2 x 20: late in the second half", HALVES, 2950, 1200),
      clockCase(
        "4 x 10: the second quarter reads 10:00",
        SHORT_QUARTERS,
        700,
        600,
      ),
      clockCase("4 x 10: inside the third quarter", SHORT_QUARTERS, 1900, 600),
      clockCase("4 x 10: the open last quarter", SHORT_QUARTERS, 2500, 600),

      parseCase("a full set, sorted by index", {
        gameId: GAME_ID,
        quarters: [
          { index: 2, startS: 1200, endS: 2100 },
          { index: 1, startS: 120, endS: 1020 },
        ],
      }),
      parseCase("an unset end is stored as null", {
        gameId: GAME_ID,
        quarters: [{ index: 1, startS: 60 }],
      }),
      parseCase("rejects a body that is not an object", "quarters"),
      parseCase("rejects a malformed game id", {
        gameId: "game-1",
        quarters: [{ index: 1, startS: 60 }],
      }),
      parseCase("rejects an empty set", { gameId: GAME_ID, quarters: [] }),
      parseCase("rejects quarters that are not a list", {
        gameId: GAME_ID,
        quarters: { index: 1, startS: 60 },
      }),
      parseCase("rejects more than four quarters", {
        gameId: GAME_ID,
        quarters: [1, 2, 3, 4, 5].map((index) => ({
          index,
          startS: index * 1000,
        })),
      }),
      parseCase(
        "two halves: a full set",
        { gameId: GAME_ID, quarters: HALVES },
        2,
      ),
      parseCase(
        "two halves: rejects a third period",
        {
          gameId: GAME_ID,
          quarters: [...HALVES, { index: 3, startS: 3100 }],
        },
        2,
      ),
      parseCase(
        "two halves: rejects an index past the count",
        { gameId: GAME_ID, quarters: [{ index: 3, startS: 60 }] },
        2,
      ),
      parseCase(
        "4 x 10: a full set",
        { gameId: GAME_ID, quarters: SHORT_QUARTERS },
        4,
      ),
      parseCase("rejects index 0", {
        gameId: GAME_ID,
        quarters: [{ index: 0, startS: 60 }],
      }),
      parseCase("rejects a fractional index", {
        gameId: GAME_ID,
        quarters: [{ index: 1.5, startS: 60 }],
      }),
      parseCase("rejects a negative start", {
        gameId: GAME_ID,
        quarters: [{ index: 1, startS: -1 }],
      }),
      parseCase("rejects an end at the start", {
        gameId: GAME_ID,
        quarters: [{ index: 1, startS: 60, endS: 60 }],
      }),
      parseCase("rejects a gap in the indices", {
        gameId: GAME_ID,
        quarters: [
          { index: 1, startS: 60 },
          { index: 3, startS: 2000 },
        ],
      }),
      parseCase("rejects a repeated index", {
        gameId: GAME_ID,
        quarters: [
          { index: 1, startS: 60 },
          { index: 1, startS: 2000 },
        ],
      }),
      parseCase("rejects a quarter starting before the previous one", {
        gameId: GAME_ID,
        quarters: [
          { index: 1, startS: 1200 },
          { index: 2, startS: 120 },
        ],
      }),
      parseCase("rejects overlapping quarters", {
        gameId: GAME_ID,
        quarters: [
          { index: 1, startS: 120, endS: 1300 },
          { index: 2, startS: 1200 },
        ],
      }),
    ],
  };
}
